test/test_tensor_ops.py

```diff
@@ -0,0 +1,411 @@
+        else:
+            dtypes = self.all_dtypes if dtypes is None else dtypes
+        super().__init__(name,
+                         dtypes=dtypes,
+                         supports_autograd=supports_autograd,
+                         check_batched_grad=False,
+                         check_batched_gradgrad=False,
+                         **kwargs)
+
+    def get_devices(self, skip_device=None, max_device_indices=2):
+        """Generator of existing device instances except skip_device.
+        """
+        if isinstance(skip_device, str):
+            skip_device = torch.device(skip_device)
+        for device_name in torch.testing.get_all_device_types():
+            if device_name == 'cuda':
+                for index in range(min(max_device_indices, torch.cuda.device_count())):
+                    device = torch.device(device_name, index)
+                    if skip_device != device:
+                        yield device
+            else:
+                device = torch.device(device_name)
+                if skip_device != device:
+                    yield device
+
+    def get_convertible_dtypes(self, dtype):
+        """Generate dtypes that allow lossless conversion from dtype and then
+        back to dtype.
+        """
+        for types in [
+                [torch.complex32, torch.complex64, torch.complex128],
+                [torch.float64, torch.complex128],
+                [torch.float32, torch.float64, torch.complex64, torch.complex128],
+                [torch.float16, torch.float32, torch.float64, torch.complex32, torch.complex64, torch.complex128],
+                [torch.bfloat16, torch.float16, torch.float32, torch.float64, torch.complex32, torch.complex64, torch.complex128],
+                [torch.int8, torch.int16, torch.int32, torch.int64],
+                [torch.bool, torch.int8, torch.int16, torch.int32, torch.int64],
+                [torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64]]:
+            if dtype in types:
+                for dt in reversed(types):
+                    yield dt
+                    if dt == dtype:
+                        return
+        assert 0, dtype
+
+
+def tensor_variants(device, dtype, requires_grad):
+    """Generator of random tensors with variable size and strides.
+    """
+    for discontiguous in [False, True]:
+        if discontiguous and dtype in [torch.bfloat16, torch.float16]:
+            # Skipping to avoid the following failures:
+            #   RuntimeError: "index_select" not implemented for 'BFloat16'
+            #   RuntimeError: "index_select" not implemented for 'Half'
+            continue
+        for size in [(5,), (3, 4)]:
+            yield make_tensor(size, device, dtype,
+                              requires_grad=requires_grad,
+                              discontiguous=discontiguous)
+
+
+def sample_inputs_tensor(op_info, device, dtype, requires_grad, pin_memory=False):
```

Comment:
This generator is so complicated it makes me wonder if pairing sample inputs with references is a good idea. I suggest separating this PR from the OpInfo changes and just adding these tests to test_tensor_creation_ops.py. My biggest concern with this pattern is that this is one "mega" test, but it seems naturally separable into more granular tests. I understand this PR pairs this giant test with a mechanism to better identify which SampleInput failed, but I'm not sure that's a perfect substitute for test granularity.